File: tasks/python3/mergeConfig.py
#!/usr/bin/env python

from util import dict_merge
from optparse import OptionParser
import os
import sys
import json

prog = os.path.basename(__file__)
base_dir = os.path.join(os.path.dirname(__file__), "..")
version = "2.0.0"
help_description = """\
Concatenates all configuration items a directory into one configuration file.
"""

# MAIN
parser = OptionParser(usage="Usage: %s <input_dir> <output_file>" % prog,
                      version="%s version %s" % (prog, version),
                      epilog=help_description)

(options, args) = parser.parse_args()
if len(args) != 2:
    parser.error("Invalid number of arguments")

input_dir = args[0]
output_file = args[1]

conf = {}
file_count = 0
for root, dirs, files in os.walk(input_dir):
    dirs.sort()
    for file in sorted(files):
        try:
            if not file.endswith(".json"):
              continue
            file_count += 1
            with open(os.path.join(root, file), "r", encoding="utf-8") as fp:
                data = json.load(fp)
            dict_merge(conf, data)
        except Exception as e:
            sys.stderr.write("ERROR: %s: %s\n" %
                             (os.path.join(root, file), str(e)))
            sys.exit(1)

json_options = {}
json_options["indent"] = 2
json_options["separators"] = (',', ': ')

with open(output_file, "w", encoding="utf-8") as fp:
    json.dump(conf, fp, **json_options)

print("%s: %s file(s) merged into %s" % (prog, file_count,
    os.path.basename(output_file)))

File: tasks/build-options/mergeConfig.js
const { dictMerge } = require('./util')
const fs = require('fs')
const path = require('path')
const yargs = require('yargs')
const prog = path.basename(__filename)

const options = yargs
  .usage('Usage: $0 [inputDir] [outputDir]')
  .option('inputDir', {
    demandOption: true,
    alias: 'i',
    type: 'string',
    description: 'getcapabilities input directory'
  })
  .option('outputFile', {
    demandOption: true,
    alias: 'o',
    type: 'string',
    description: 'wmts output directory'
  })
  .epilog('Concatenates all configuration items a directory into one configuration file.')

const { argv } = options
if (!argv.inputDir && !argv.outputFile) {
  throw new Error('Invalid number of arguments')
}

const conf = {}
let fileCount = 0

async function main () {
  const inputDir = argv.inputDir
  const outputFile = argv.outputFile
  await mergeFiles(inputDir)

  fs.writeFileSync(outputFile, JSON.stringify(conf, null, 2), 'utf-8')

  console.warn(`${prog}: ${fileCount} file(s) merged into ${path.basename(outputFile)}`)
}

async function mergeFiles (inputDir) {
  const files = fs.readdirSync(inputDir)
  for (const file of files) {
    try {
      if (file.endsWith('.json')) {
        fileCount += 1
        const data = JSON.parse(fs.readFileSync(path.join(inputDir, file), 'utf-8'))
        await dictMerge(conf, data)
      } else if (fs.existsSync(path.join(inputDir, file)) && fs.lstatSync(path.join(inputDir, file)).isDirectory()) {
        subDir = path.join(inputDir, file)
        await mergeFiles(subDir)
      }
    } catch (error) {
      throw new Error(`ERROR: ${path.join(inputDir, file)}: ${error.message}`)
    }
  }
}

main().catch((err) => {
  console.error(err.stack)
})
